### ISSUE TITLE
Support await'ing a Task without throwing

### ISSUE THREAD

**Issue by stephentoub** [dotnet/runtime#22144]
EDITED 5/24/2023 by @stephentoub:
Latest proposal at https://github.com/dotnet/runtime/issues/22144#issuecomment-1561983918
```diff
namespace System.Threading.Tasks;

public class Task
{
+    public ConfiguredTaskAwaitable ConfigureAwait(ConfigureAwaitOptions options);
}

+[Flags]
+public enum ConfigureAwaitOptions
+{
+    None = 0,
+    ContinueOnCapturedContext = 0x1,
+    SuppressExceptions = 0x2,
+    ForceYielding = 0x4,
+    ForceAsynchronousContinuation = 0x8,
+}
```

(I'm not sure yet if we actually need to ship ForceAsynchronousContinuation now.  We might hold off if we don't have direct need in our own uses.)

---

**EDIT**: See https://github.com/dotnet/runtime/issues/22144#issuecomment-943445439 for an up-to-date API proposal.

```diff
namespace System.Threading.Tasks;

+[Flags]
+public enum TaskAwaitBehavior
+{
+    Default = 0x0,
+    NoContinueOnCapturedContext = 0x1,
+    NoThrow = 0x2,
+}

public partial class Task
{
     public ConfiguredTaskAwaitable ConfigureAwait(bool continueOnCapturedContext);
+    public ConfiguredTaskAwaitable ConfigureAwait(TaskAwaitBehavior awaitBehavior);
}

public partial class Task<TResult>
{
     public new ConfiguredTaskAwaitable<TResult> ConfigureAwait(bool continueOnCapturedContext);
+    public new ConfiguredTaskAwaitable<TResult> ConfigureAwait(TaskAwaitBehavior awaitBehavior);
}

public partial struct ValueTask
{
     public ConfiguredValueTaskAwaitable ConfigureAwait(bool continueOnCapturedContext);
+    public ConfiguredValueTaskAwaitable ConfigureAwait(TaskAwaitBehavior awaitBehavior);
}

public partial struct ValueTask<TResult>
{
     public ConfiguredTaskAwaitable<TResult> ConfigureAwait(bool continueOnCapturedContext);
+    public ConfiguredTaskAwaitable<TResult> ConfigureAwait(TaskAwaitBehavior awaitBehavior);
}
```

<details>
<summary>Original post</summary>
Currently there isn't a great way to await a Task without throwing (if the task may have faulted or been canceled).  You can simply eat all exceptions:

```C#
try { await task; } catch { }
```
but that incurs the cost of the throw and also triggers first-chance exception handling.  You can use a continuation:
```C#
await task.ContinueWith(delegate { }, CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
```
but that incurs the cost of creating and running an extra task.  The best way in terms of run-time overhead is to use a custom awaiter that has a nop GetResult:
```C#
internal struct NoThrowAwaiter : ICriticalNotifyCompletion
{
    private readonly Task _task;
    public NoThrowAwaiter(Task task) { _task = task; }
    public NoThrowAwaiter GetAwaiter() => this;
    public bool IsCompleted => _task.IsCompleted;
    public void GetResult() { }
    public void OnCompleted(Action continuation) => _task.GetAwaiter().OnCompleted(continuation);
    public void UnsafeOnCompleted(Action continuation) => OnCompleted(continuation);
}
...
await new NoThrowAwaiter(task);
```
but that's obviously more code than is desirable.  It'd be nice if functionality similar to that last example was built-in.

**Proposal**
Add a new overload of `ConfigureAwait`, to both `Task` and `Task<T>`.  Whereas the current overload accepts a `bool`, the new overload would accept a new `ConfigureAwaitBehavior` enum:
```C#
namespace System.Threading.Tasks
{
    [Flags]
    public enum ConfigureAwaitBehavior
    {
        NoCapturedContext = 0x1, // equivalent to ConfigureAwait(false)
        NoThrow = 0x2, // when set, no exceptions will be thrown for Faulted/Canceled
        Asynchronous = 0x4, // force the continuation to be asynchronous
        ... // other options we might want in the future
    }
}
```
Then with `ConfigureAwait` overloads:
```C#
namespace System.Threading.Tasks
{
    public class Task
    {
        ...
        public ConfiguredTaskAwaitable ConfigureAwait(ConfigureAwaitBehavior behavior);
    }

    public class Task<TResult> : Task
    {
        ...
        public ConfiguredTaskAwaitable<TResult> ConfigureAwait(ConfigureAwaitBehavior behavior);
    }
}
```
code that wants to await without throwing can write:
```C#
await task.ConfigureAwait(ConfigureAwaitBehavior.NoThrow);
```
or that wants to have the equivalent of `ConfigureAwait(false)` and also not throw:
```C#
await task.ConfigureAwait(ConfigureAwaitBehavior.NoCapturedContext | ConfigureAwaitBehavior.NoThrow);
```
etc.

From an implementation perspective, this will mean adding a small amount of logic to ConfiguredTaskAwaiter, so there's a small chance it could have a negative imp

**Alternatives**
An alternative would be to add a dedicated API like `NoThrow` to `Task`, either as an instance or as an extension method, e.g.
```C#
await task.NoThrow();
```
That however doesn't compose well with wanting to use `ConfigureAwait(false)`, and we'd likely end up needing to add a full matrix of options and supporting awaitable/awaiter types to enable that.

Another option would be to add methods like NoThrow to ConfiguredTaskAwaitable, so you could write:
```C#
await task.ConfigureAwait(true).NoThrow();
```
etc.

And of course an alternative is to continue doing nothing and developers that need this can write their own awaiter like I did earlier.
</details>

**Comment by davidfowl** [comment]
Where did we land on this API wise? Do we like the ConfigureAwait approach or the NoThrow extension method (I prefer that).

**Comment by stephentoub** [comment]
I prefer the ConfigureAwait approach, e.g. `await task.ConfigureAwait(ConfigureAwaitBehavior.NoThrow | ConfigureAwaitBehavior.NoCapturedContext);`, rather than having lots of different methods for all the various ways we might want to support awaiting and that all need to compose in some fashion (this is why the method is called ConfigureAwait today, because the idea was it could be extended in this manner in the future).  This also ties in with #27723, for example, where we'd have an overload that would let you do `await task.ConfigureAwait(timeout)`, an overload that would let you pass a cancellation token, an overload that would let you pass a timeout and a cancellation token and a ConfigureAwaitBehavior, etc.  It all ends up just being about configuring the await.
#32030

@eiriktsarpalis, are you still driving the design of this?

**Comment by eiriktsarpalis** [comment]
I started working on a prototype a few months back, but it got dropped on the floor. I'll see if I can revive it.

**Comment by ReubenBond** [comment]
A developer might want one of several behaviors:

1. Signal completion but swallow exceptions
2. Signal completion and return the result of the task but swallow exceptions and return `default` if the task was canceled/faulted
3. For either: Ensure that `TaskScheduler.UnobservedTaskExeption` does not fire if this task's failure is not observed
4. Signal completion, exceptions, and cancellation without throwing for the latter two.

To illustrate, if a developer writes `var result = await task.ConfigureAwait(ConfigureAwaitBehavior.NoThrow)` where `task` is `Task<string>`, what would `result` be in the case that `task` failed? I suppose `result` may be `null` since we can't dependently type the return value based on the `ConfigureAwaitBehavior` value. In the case of `await task.NoThrow()`, it could be decided that it acts like `Task` and has no result value.

In Orleans, we typically want behavior 3 (not to invalidate the other two).

If we had an extension method similar to Rx's [`Observable.Materialize<TSource>` method](https://docs.microsoft.com/en-us/previous-versions/dotnet/reactive-extensions/hh229453(v=vs.103)) then we could gracefully await completion, exceptions, or results without throwing. Eg calling it `.AsNotification()` for exposition purposes:

``` C#
Result<T> result = await task.AsNotification();

return result switch
{
 (Canceled, _, _) => null, // Ignore if the task was canceled.
 (Faulted, _, ex) => await task, // throw if the task faulted
 (Completed, value, _) => value,
  _ => throw new NotSupportedException("...")
};

// Given deconstruction, it could be:
var (status, value, exception) = await task.AsNotification();
```

That doesn't satisfy behavior 3, but it satisfies behavior 1, 2, and 4, I believe.

**Comment by stephentoub** [comment]
> if a developer writes var result = await task.ConfigureAwait(ConfigureAwaitBehavior.NoThrow) where task is Task<string>, what would result be in the case that task failed?

It would be `default(T)`.  With NoThrow, you have to check the status of `task` afterwards to know whether it was successful, faulted, or canceled, and you'd only use the result if it was valid.

**Comment by sharwell** [comment]
> I prefer the ConfigureAwait approach

I also prefer this approach. Considering how clean it is, even if it starts out slower in some sense I'd be interested to rule out JIT improvements targeting common patterns before letting it impact the API.

**Comment by ReubenBond** [comment]
If a developer writes `_ = task.ConfigureAwait(ConfigureAwaitBehavior.NoThrow)` without awaiting, would an exception still bubble up to `TaskScheduler.UnobservedTaskException`?

**Comment by stephentoub** [comment]
> would an exception still bubble up to TaskScheduler.UnobservedTaskException?

We can decide, but I think the answer should be "yes".

**Comment by sharwell** [comment]
> would an exception still bubble up to `TaskScheduler.UnobservedTaskException`

My instinct here is _yes_. Normally the `await` operation observes the exception for a faulted task, but `ConfigureAwaitBehavior.NoThrow` means it does not. Code is expected to check the task status separately in this case, and failure to do so would result in an `UnobservedTaskException`.

**Comment by stephentoub** [comment]
> If a developer writes _ = task.ConfigureAwait(ConfigureAwaitBehavior.NoThrow) without awaiting

Oh, wait, "without awaiting"?  If there's no await, the above is entirely a nop.

**Comment by willdean** [comment]
A little bit of bike-shedding, but should the enum be `AwaitBehavior` rather than `ConfigureAwaitBehaviour`?

After all, it's describing the behaviour of the await, not the behaviour of the the `ConfigureAwait` function - the latter is just there to configure the behaviour.

**Comment by Clockwork-Muse** [comment]
.... Side note: If we ever get a `Result<T, TError>` union, are we still anticipating needing this?  The intent being that exceptions would very much be "fatal" errors, whereas "ignore the small stuff" would be handled by returning the `Result` union.

**Comment by GSPP** [comment]
> > would an exception still bubble up to TaskScheduler.UnobservedTaskException?
> 
> We can decide, but I think the answer should be "yes". If someone doesn't want that, they just check task.Exception, just as they can today.

My intuition would be "no" because the developer already consciously attended to this task and decided in his head how he wants to deal with it. When `UnobservedTaskException` was introduced, it was meant as a debugging aid to flag likely bugs where important things are dropped on the floor. If a task is awaited I'd see this as a strong heuristic sign that the developer dealt with it.

**Comment by stephentoub** [comment]
> My intuition would be "no" because the developer already consciously attended to this task and decided in his head how he wants to deal with it.

I don't believe we know that just from being told the await shouldn't throw.  In my experience, such a request often comes from wanting to join and then handle failures in some special way and not wanting to pay the cost of a throw/catch in order to handle it.  On top of that, the exception really isn't being observed, so it's arguably also more correct to not treat the exception as observed.

**Comment by eiriktsarpalis** [comment]
> > if a developer writes var result = await task.ConfigureAwait(ConfigureAwaitBehavior.NoThrow) where task is Task, what would result be in the case that task failed?
>
> It would be default(T).

I'm not particularly comfortable with that approach. In most cases it would violate user invariants and is unsound if the result is a non-nullable reference type. I think we should either a) restrict non-throwing awaitables to non-generic tasks only or b) use a dedicated awaiter type whose `GetResult` method explicitly returns `TResult?`.

> Asynchronous = 0x4, // force the continuation to be asynchronous

I'm almost certain this will require some form of state machine support, since current SM's will check if the awaiter has completed before attempting to subscribe their continuation. While we could force asynchrony by having the awaiter's `IsCompleted` property always return false, this feels like a hack to me.

**Comment by stephentoub** [comment]
> force asynchrony by having the awaiter's IsCompleted property always return false, this feels like a hack to me.

That's exactly what it would do.  I don't see why it's a hack; that's the purpose of the IsCompleted property.

> I think we should either a) restrict non-throwing awaitables to non-generic tasks only or b) use a dedicated awaiter type whose GetResult method explicitly returns TResult?.

I personally think both of those are worse, especially when factoring in composability with other options on the await.

**Comment by eiriktsarpalis** [comment]
> > force asynchrony by having the awaiter's IsCompleted property always return false, this feels like a hack to me.
>
> That's exactly what it would do. I don't see why it's a hack; that's the purpose of the IsCompleted property.

I mean, strictly speaking it is pretending to not be completed so that the state machine is forced to pass its continuation. I'd probably be less concerned if this weren't a public property.

> Also, I don't understand what `TResult?` would mean:

I was motivated by the assumption that returning `default` for reference types is more harmful than value types. But then again value types can contain references so probably not that important. Alternatively, we could perhaps consider adding an analyzer/fixer that detects expressions of the form `Foo foo = await task.ConfigureAwait(AwaitBehavior.NoThrow);` and recommends changing those to `Foo? foo = await task.ConfigureAwait(AwaitBehavior.NoThrow);`.

**Comment by stephentoub** [comment]
> strictly speaking it is pretending to not be completed

The definition of IsCompleted is up to the awaiter.  In this case, that definition includes not supporting synchronous completion.  It's also in the System.Runtime.CompilerServices namespace because it's part of the async/await infrastructure and meant to be consumed by the compiler via a very specific pattern.  IsCompleted is a simple, understandable name, but if it helps you sleep better :-), think of it as "StateMachineShouldProceedWithoutHookingUpACallback".

> Alternatively, we could perhaps consider adding an analyzer

Sounds reasonable.

**Comment by sharwell** [comment]
> I mean, strictly speaking it is pretending to not be completed so that the state machine is forced to pass its continuation. I'd probably be less concerned if this weren't a public property.

This is how `Task.Yield()` is implemented:

https://github.com/dotnet/runtime/blob/8a52f1e948b6f22f418817ec1068f07b8dae2aa5/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/YieldAwaitable.cs#L50

**Comment by stephentoub** [comment]
> This is how Task.Yield() is implemented

Exactly

**Comment by eiriktsarpalis** [comment]
I've knocked together [a (far from complete) prototype](https://github.com/eiriktsarpalis/runtime/commit/6b0487c228e65f49f3f5b45a3a236e60e1e70ae7) that combines the features of this issue as well as #27723 (for heap-allocated tasks only). 

Turns out it is possible to extend the existing `ConfiguredTaskAwaitable` structs so that they store a `ConfigureAwaitBehavior` enum instead of the current `continueOnCapturedContext` boolean without breaking changes or significantly impacting the hot path for async methods. Running the [existing async method benchmarks](https://github.com/dotnet/performance/blob/master/src/benchmarks/micro/libraries/System.Threading.Tasks/Perf.AsyncMethods.cs) seems to confirm that there is no observable perf regression when compared with the current main branch.

From an API perspective the prototype brings the following changes:
```csharp
namespace System.Runtime.CompilerServices
{
    public readonly struct CancelableConfiguredTaskAwaitable
    {
        public System.Runtime.CompilerServices.CancelableConfiguredTaskAwaitable.CancelableConfiguredTaskAwaiter GetAwaiter() { throw null; }
        public readonly struct CancelableConfiguredTaskAwaiter : System.Runtime.CompilerServices.ICriticalNotifyCompletion, System.Runtime.CompilerServices.INotifyCompletion
        {
            public bool IsCompleted { get { throw null; } }
            public void GetResult() { }
            public void OnCompleted(System.Action continuation) { }
            public void UnsafeOnCompleted(System.Action continuation) { }
        }
    }

    public readonly struct CancelableConfiguredTaskAwaitable<TResult>
    {
        public System.Runtime.CompilerServices.CancelableConfiguredTaskAwaitable<TResult>.CancelableConfiguredTaskAwaiter GetAwaiter() { throw null; }
        public readonly partial struct CancelableConfiguredTaskAwaiter : System.Runtime.CompilerServices.ICriticalNotifyCompletion, System.Runtime.CompilerServices.INotifyCompletion
        {
            public bool IsCompleted { get { throw null; } }
            public TResult GetResult() { throw null; }
            public void OnCompleted(System.Action continuation) { }
            public void UnsafeOnCompleted(System.Action continuation) { }
        }
    }
}

namespace System.Threading.Tasks
{
    [Flags]
    public enum ConfigureAwaitBehavior
    {
        Default = 0x0,
        NoCapturedContext = 0x1,
        NoThrow = 0x2,
        Asynchronous = 0x4,
    }

    public partial class Task
    {
        public System.Runtime.CompilerServices.ConfiguredTaskAwaitable ConfigureAwait(ConfigureAwaitBehavior awaitBehavior) { throw null; }
        public new System.Runtime.CompilerServices.CancelableConfiguredTaskAwaitable ConfigureAwait(bool continueOnCapturedContext, System.Threading.CancellationToken cancellationToken) { throw null; }
        public new System.Runtime.CompilerServices.CancelableConfiguredTaskAwaitable ConfigureAwait(ConfigureAwaitBehavior awaitBehavior, System.Threading.CancellationToken cancellationToken) { throw null; }
    }

    public partial class Task<TResult>
    {
        public new System.Runtime.CompilerServices.ConfiguredTaskAwaitable<TResult> ConfigureAwait(ConfigureAwaitBehavior awaitBehavior) { throw null; }
        public new System.Runtime.CompilerServices.CancelableConfiguredTaskAwaitable<TResult> ConfigureAwait(bool continueOnCapturedContext, System.Threading.CancellationToken cancellationToken) { throw null; }
        public new System.Runtime.CompilerServices.CancelableConfiguredTaskAwaitable<TResult> ConfigureAwait(ConfigureAwaitBehavior awaitBehavior, System.Threading.CancellationToken cancellationToken) { throw null; }
    }
}
```

PTAL. If you feel this is the right design I can maybe file it as an API proposal (which also includes equivalent ValueTask extensions).

**Comment by davidfowl** [comment]
Can you add some samples to this?

**Comment by stephentoub** [comment]
> PTAL. If you feel this is the right design I can maybe file it as an API proposal (which also includes equivalent ValueTask extensions).

It generally looks good.  You can just edit the top part of this issue.  A few things:
- i think we could get away without the bool,CT overloads.
- Asynchronous => ForceAsynchronous, or something that makes it more clear what that's doing
- We'll want the same set for ValueTask, as you noted. 
- I think there should be overloads taking a TimeSpan timeout; it could just be on the longest overload of each set.  While it can be done using the token,  i suspect some folks doing so will end up doing `new CTS(timeout).Token` and never disposing it.

**Comment by davidfowl** [comment]
> I think there should be overloads taking a TimeSpan timeout; it could just be on the longest overload of each set. While it can be done using the token, i suspect some folks doing so will end up doing new CTS(timeout).Token and never disposing it.

Imagine if we had an analyzer 😄

**Comment by stephentoub** [comment]
> Turns out it is possible to extend the existing ConfiguredTaskAwaitable structs so that they store a ConfigureAwaitBehavior enum instead of the current continueOnCapturedContext boolean without breaking changes or significantly impacting the hot path for async methods.

I'm a little surprised about that.  I'd have expected the Asynchronous flag to need to be checked in IsCompleted and the NoThrow flag to need to be checked in GetResult, resulting in several more branches on the hot path as well as either more code that gets inlined into lots of call sites or methods that once were inlined no longer being so.

> Running the existing async method benchmarks seems to confirm that there is no observable perf regression when compared with the current main branch.

Those don't use ConfigureAwait.

**Comment by theodorzoulias** [comment]
Does it make any sense to add another value to the `ConfigureAwaitBehavior` enum, the `CaptureCurrentTaskScheduler = 0x8` value? That value would change the behavior of the `await`, so that it captures the `TaskScheduler.Current` instead of the `SynchronizationContext.Current`. I see a potential for some cool `TaskScheduler`s with this option. I could open a separate issue with this suggestion, unless it is out of the question for being technically infeasible.

**Comment by stephentoub** [comment]
> Does it make any sense to add another value to the ConfigureAwaitBehavior enum, the CaptureCurrentTaskScheduler = 0x8 value?

If someone wanted to do this,  they could set the sync ctx to null,  then call the method,  then reset the sync ctx. The fact that I've literally never seen anyone do that for the purpose of preferring the current TaskScheduler suggests to me its not important enough to add API for. What are the scenarios you have in mind?

**Comment by eiriktsarpalis** [comment]
>>Turns out it is possible to extend the existing ConfiguredTaskAwaitable structs so that they store a ConfigureAwaitBehavior enum instead of the current continueOnCapturedContext boolean without breaking changes or significantly impacting the hot path for async methods.
>
>I'm a little surprised about that. I'd have expected the Asynchronous flag to need to be checked in IsCompleted and the NoThrow flag to need to be checked in GetResult, resulting in several more branches on the hot path as well as either more code that gets inlined into lots of call sites or methods that once were inlined no longer being so.

You're right about [`IsCompleted`](https://github.com/eiriktsarpalis/runtime/commit/6b0487c228e65f49f3f5b45a3a236e60e1e70ae7#diff-ab182f77ea6534b106fb3c228f146601ea88fe8cda6fd713dbe5559c3c7c5f3cR496-R498), as far as `GetResult` is concerned it is [possible to avoid any checks in the hot path](https://github.com/eiriktsarpalis/runtime/commit/6b0487c228e65f49f3f5b45a3a236e60e1e70ae7#diff-ab182f77ea6534b106fb3c228f146601ea88fe8cda6fd713dbe5559c3c7c5f3cR615-R620) since task internals are available to the awaiter and any additional checks would only be needed in the case where the task is [known to have completed with an exception](https://github.com/eiriktsarpalis/runtime/commit/6b0487c228e65f49f3f5b45a3a236e60e1e70ae7#diff-ab182f77ea6534b106fb3c228f146601ea88fe8cda6fd713dbe5559c3c7c5f3cR215-R219).

> Running the existing async method benchmarks seems to confirm that there is no observable perf regression when compared with the current main branch.
>
> > Those don't use ConfigureAwait.

Good point, I re-ran the benchmarks using `ConfigureAwait(false)` where possible but still couldn't detect any discernible regressions, but probably that is expected given the nature of the particular microbenchmarks.

**Comment by theodorzoulias** [comment]
> If someone wanted to do this, they could set the sync ctx to null, then call the method, then reset the sync ctx. The fact that I've literally never seen anyone do that for the purpose of preferring the current TaskScheduler suggests to me its not important enough to add API for. What are the scenarios you have in mind?

I may open a new issue about this suggestion, to avoid polluting this thread with something that may just be a bad idea.

**Comment by eiriktsarpalis** [comment]
> You're right about IsCompleted, as far as GetResult is concerned it is possible to avoid any checks in the hot path since task internals are available to the awaiter and any additional checks would only be needed in the case where the task is known to have completed with an exception.

That being said, it is reasonable to assume that future additions to the `ConfigureAwaitBehavior` enum might be not be implementable due to performance considerations in the hot path. Perhaps the best way forward is to only allow modifying the await behaviour in the new cancellable awaitable:
```csharp
public CancelableConfiguredTaskAwaitable ConfigureAwait(this Task task, ConfigureAwaitBehavior awaitBehavior) => 
            new CancelableConfiguredTaskAwaitable(task, awaitBehavior, default(CancellationToken));
```
But this begs the question what the name for the new awaitable should be. Clearly, it does more than just being a cancellable awaitable.

**Comment by eiriktsarpalis** [comment]
One potential issue I identified when working on the `ValueTask` prototype is that the current `ConfiguredValueTaskAwaiter.GetResult()` implementation delegates to [`IValueTaskSource.GetResult()`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.sources.ivaluetasksource.getresult?view=net-5.0#System_Threading_Tasks_Sources_IValueTaskSource_GetResult_System_Int16_): clearly we cannot make `IValueTaskSource` accept `ConfigureAwaiteBehavior`-like arguments without breaking changes, so in order to support `NoThrow` we would probably need to change the new awaiter behaviour so that the result is polled using [`IValueTaskSource.GetStatus`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.sources.ivaluetasksource.getresult?view=net-5.0#System_Threading_Tasks_Sources_IValueTaskSource_GetResult_System_Int16_) calls. Depending on the underlying `IValueTaskSource` implementation, this could result on subtle changes depending on the specified await behavior.

**Comment by stephentoub** [comment]
> So, in order to support NoThrow we would probably need to change the new awaiter behaviour so that the result is polled using IValueTaskSource.GetStatus calls. Depending on the underlying IValueTaskSource implementation, this could result on subtle changes depending on the specified await behavior.

Yup. That's fine. An IValueTaskSource implementation needs to support this pattern of access anyway, as code sometimes does fast-path checks that involves checking members of ValueTask like IsFaulted.  And as for a behavior change, even if it was perceptable, code would be opting in by utilizing the new NoThrow option.

**Comment by stephentoub** [comment]
> But this begs the question what the name for the new awaitable should be. Clearly, it does more than just being a cancellable awaitable.

While we should give some thought to a name so that it's not abysmal, the name here also isn't a big deal: it should almost never show up in user code, and is buried in the System.Runtime.CompilerServices namespace (which is not known for great type names, because they're not intended to be spelled out in code).

**Comment by eiriktsarpalis** [comment]
>> So, in order to support NoThrow we would probably need to change the new awaiter behaviour so that the result is polled using IValueTaskSource.GetStatus calls. Depending on the underlying IValueTaskSource implementation, this could result on subtle changes depending on the specified await behavior.
>
> Yup. That's fine. An IValueTaskSource implementation needs to support this pattern of access anyway, as code sometimes does fast-path checks that involves checking members of ValueTask like IsFaulted. And as for a behavior change, even if it was perceptable, code would be opting in by utilizing the new NoThrow option.

There is a corner case in which we would still need [eat exceptions](https://github.com/eiriktsarpalis/runtime/blob/6fa0baf1388ebca6ac5b2b7da67859aeb12d95cb/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/ConfiguredCancelableValueTaskAwaitable.cs#L340-L352), but that only concerns calling `awaiter.GetResult()` in cases where the task operation is still pending; this shouldn't impact state machines.

> While we should give some thought to a name so that it's not abysmal, the name here also isn't a big deal

I think `ConfiguredCancelableValueTaskAwaitable` is a good compromise in this case.

**Comment by stephentoub** [comment]
> There is a corner case in which we would still need eat exceptions, but that only concerns calling awaiter.GetResult() in cases where the task operation is still pending; this shouldn't impact state machines.

That's erroneous usage, anyway, so I'm not bothered by it. I think I'd actually prefer leaving out the try/catch, as a good IValueTaskSource implementation will throw for such misuse, and I wouldn't want the try/catch to eat that nor to prevent inlining if it would otherwise happen.

**Comment by pentp** [comment]
Please consider a shorter enum name that doesn't start with a verb, e.g. what @willdean proposed (`AwaitBehavior`). `ConfigureAwait(ConfigureAwaitBehavior.NoThrow)` is a lot of redundancy.

**Comment by eiriktsarpalis** [comment]
I've created issue #47525 which combines the proposals of #27723 and #22144. Closing this one.

**Comment by eiriktsarpalis** [comment]
Reopening since we ended up cutting "SuppressExceptions" support in #47525.

**Comment by theodorzoulias** [comment]
I hope that this will make it to .NET 8!

**Comment by theodorzoulias** [comment]
After watching the [video](https://www.youtube.com/watch?v=etwMhIKj0tI&t=0h35m3s), I would like to suggest an alternative API with narrower scope, that doesn't mess with the nullability of the `TResult`. The idea is that if you want the `await` to not throw an exception, you agree to lose the result of the awaitable. So the configured awaitable is always non-generic, for both generic and non-generic awaitables.

### API Proposal

```C#
namespace System.Runtime.CompilerServices
{
    public struct ConfiguredTaskAwaitable
    {
        public static ConfiguredTaskAwaitable NoThrow();
    }
    public struct ConfiguredTaskAwaitable<TResult>
    {
        public static ConfiguredTaskAwaitable NoThrow();
    }
    public struct ConfiguredValueTaskAwaitable
    {
        public static ConfiguredValueTaskAwaitable NoThrow();
    }
    public struct ConfiguredValueTaskAwaitable<TResult>
    {
        public static ConfiguredValueTaskAwaitable NoThrow();
    }
}
```

The `NoThrow` configuration is applied only on a configured awaitable, not directly on a `Task` or a `ValueTask`. So the code below will not compile:

```C#
await Task.Delay(1000, token).NoThrow();
```

...but this will be possible instead:

```C#
await Task.Delay(1000, token).ConfigureAwait(true).NoThrow();
```

This way the `NoThrow` functionality will be less exposed, but nevertheless it will be available for anyone who needs it.

### API Usage

1. Writing a method that invokes a delegate in periodic fashion, and accepts a `CancellationToken` with stopping semantics:

```C#
static async Task PeriodicAsync(Func<Task> action,
    TimeSpan interval,
    CancellationToken stoppingToken = default)
{
    using var timer = new PeriodicTimer(interval);
    while (!stoppingToken.IsCancellationRequested)
    {
        await action();
        await timer.WaitForNextTickAsync(stoppingToken).ConfigureAwait(true).NoThrow();
    }
}
```

2. Writing a variant of the [`SelectAwait`](https://github.com/dotnet/reactive/blob/85f1eb7c53e27cccdbeee3e0b044916168843fcc/Ix.NET/Source/System.Linq.Async/System/Linq/Operators/Select.cs#L81) LINQ operator (System.Linq.Async), that allows concurrency for each pair of iteration/projection operations. In case the consumer of this API abandons the enumeration prematurely, we want to dispose the source enumerator. We can't dispose it before awaiting the pending `MoveNextAsync` operation, otherwise the `DisposeAsync` might throw. But we don't really want to propagate a possible exception of the `MoveNextAsync`. The consumer has already abandon the enumeration and is now disposing our enumerator, and throwing exceptions on the `DisposeAsync` [is not recommended](https://stackoverflow.com/questions/577607/should-you-implement-idisposable-dispose-so-that-it-never-throws/577671#577671). Here comes the need for the `NoThrow` API:

```C#
static async IAsyncEnumerable<TResult> SelectAwaitConcurrent<TSource, TResult>(
    this IAsyncEnumerable<TSource> source,
    Func<TSource, ValueTask<TResult>> selector)
{
    IAsyncEnumerator<TSource> enumerator = source.GetAsyncEnumerator();
    ValueTask<bool> moveTask = enumerator.MoveNextAsync().Preserve();
    try
    {
        while (await moveTask.ConfigureAwait(false))
        {
            ValueTask<TResult> selectorTask = selector(enumerator.Current);
            moveTask = enumerator.MoveNextAsync().Preserve();
            yield return await selectorTask.ConfigureAwait(false);
        }
    }
    finally
    {
        await moveTask.ConfigureAwait(false).NoThrow(); // API usage
        await enumerator.DisposeAsync().ConfigureAwait(false);
    }
}
```

**Comment by pentp** [comment]
I believe the options that return a `ConfiguredTaskAwaitable<TResult>` (or similar) were already dismissed as too risky when the task fails.
There is one very specific implementation in dotnet/runtime: https://github.com/dotnet/runtime/blob/62e87b43207a17caa7b3d3260c1f58f155f5284e/src/libraries/System.Private.CoreLib/src/System/Threading/SemaphoreSlim.cs#L738
For general use I propose just returning the task as the result of the awaitable:
```cs
/// <summary>
/// Returns an awaitable that does not throw exceptions.
/// The caller is responsible for observing any exceptions.
/// </summary>
public static NoThrowAwaiter<T> NoThrow<T>(this T task) where T : Task => new(task);

public readonly struct NoThrowAwaiter<T> : ICriticalNotifyCompletion where T : Task
{
    private readonly T _task;
    internal NoThrowAwaiter(T task) => _task = task;
    public NoThrowAwaiter<T> GetAwaiter() => this;
    public bool IsCompleted => _task.IsCompleted;
    public void OnCompleted(Action action) => _task.GetAwaiter().OnCompleted(action);
    public void UnsafeOnCompleted(Action action) => _task.GetAwaiter().UnsafeOnCompleted(action);
    public T GetResult() => _task;
}
```
This has the additional benefit of not needing to store the task in a captured variable (it's instead only temporarily stored in the awaiter variable, which is anyway needed). A different overload+struct could add `continueOnCapturedContext` parameter support also.

**Comment by stephentoub** [comment]
> For general use I propose just returning the task as the result of the awaitable...
> The idea is that if you want the await to not throw an exception, you agree to lose the result of the awaitable...

Both approaches could be viable for Task.  Neither is appropriate for ValueTask, which can't be awaited multiple times.

**Comment by theodorzoulias** [comment]
@stephentoub what about returning a new completed `ValueTask<TResult>`, that is not the original awaited `ValueTask<TResult>`, but a [preserved](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask-1.preserve) version of it? A preserved value-task can be used with the same liberty as a reference-task.

```C#
public struct ConfiguredValueTaskAwaitable<TResult>
{
    public static ConfiguredValueTaskAwaitable<ValueTask<TResult>> NoThrow();
}
```

**Comment by stephentoub** [comment]
That's possible, but it adds cost to a method whose whole purpose is to avoid cost.

**Comment by theodorzoulias** [comment]
@stephentoub yea, it seems that an `IValueTaskSource<TResult>` has no other way of communicating an exception other than throwing from the [`GetResult`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.sources.ivaluetasksource-1.getresult) method. So the only information that can be obtained cheaply is the status of the completed `ValueTask<TResult>`, and (in case of success) the result. That's unfortunate.

**Comment by stephentoub** [comment]
I'm not overly concerned about the overhead in the case of a faulted task. I'm more concerned about the overhead in the case of a successful one. Preserving a ValueTask backed by an IValueTaskSource requires an allocation.

**Comment by theodorzoulias** [comment]
@stephentoub why so? I am expecting for the happy path to just return a completed `ValueTask<TResult>` backed by a result, not by a `Task<TResult>`. The overhead in case of failure matters, because the whole purpose is to avoid throwing and catching. If the awaiter has a `try`/`catch` internally, we are losing everything that we expected to gain.

**Comment by stephentoub** [comment]
> I am expecting for the happy path to just return a completed ValueTask<TResult> backed by a result, not by a Task<TResult>

In your initial proposal, no value is returned, so I'm not sure what this means.

If you mean for Pent's proposal, constructing a ValueTask using a result after the await is different from using Preserve as you'd suggested. Using Preserve as you suggested would need to be done before not after the continuation is hooked up. A new ValueTask could be constructed from the result, though, yes.

> The overhead in case of failure matters, because the whole purpose is to avoid throwing and catching.

I'm not saying it doesn't matter, but to clarify, the "whole purpose is to avoid throwing and catching" _if_ there's an exception... but it should still be the exceptional case that there is an exception, such that adding allocation to the success path is not a good tradeoff.

>  If the awaiter has a try/catch internally, we are losing everything that we expected to gain.

It would need one for some ValueTasks, there's no other option. But you'd only pay for that throw/catch if the ValueTask is backed by a faulting IValueTaskSource, not always.

**Comment by theodorzoulias** [comment]
@stephentoub yep, I was talking Pent's proposal, which sounded good until the realization that it requires an internal `try`/`catch` in some cases that are out of my control. Personally I would be satisfied with receiving no information whatsoever after a non-throwing `await`. This covers 90% of my use cases. And I would be quite unhappy if my `NoThrow` awaits had to pay a `try`/`catch` cost behind the scenes, in order to bring me information that I rarely need.

Regarding returning a preserved value-task, I wasn't talking literally about calling [`Preserve`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask-1.preserve) on the incomplete value-task. I meant about the effect that the `Preserve` method has: returns a different value-task that can be awaited multiple times. Sorry for being unclear.

**Comment by stephentoub** [comment]
> Personally I would be satisfied with receiving no information whatsoever after a non-throwing await. This covers 90% of my use cases.

I'm surprised by that. It's rare in my experience to care that an operation has completed but to also not care whether it was successful, what result it produced if any, or what error it encountered if any. Such scenarios certainly exist, but in my experience they're very far from the majority. I don't believe we should ship an API for this that has such constraints.

**Comment by theodorzoulias** [comment]
@stephentoub in a [previous comment](https://github.com/dotnet/runtime/issues/22144#issuecomment-1328128898) I demonstrated three of my use cases where the status/result of the asynchronous operation is irrelevant. Here is another one. Writing a many-to-one completion propagator for the TPL Dataflow library:

```C#
static void PropagateCompletion(IDataflowBlock[] sources, IDataflowBlock target)
{
    Task allSourcesCompletion = Task.WhenAll(sources.Select(s => s.Completion));
    ThreadPool.QueueUserWorkItem(async _ =>
    {
        await allSourcesCompletion.ConfigureAwait(false).NoThrow();

        Exception exception = allSourcesCompletion.IsFaulted ?
            allSourcesCompletion.Exception : null;

        if (exception is null) target.Complete(); else target.Fault(exception);
    });
}
```

Most of my use cases for the `NoThrow` are like these. I don't know how other people would like to use this API. I would be interested to see the uses cases of other people.

**Comment by stephentoub** [comment]
> where the status/result of the asynchronous operation is irrelevant. Here is another one

I don't understand... in that example you _are_ accessing the task after the await... you _do_ care about details of its completion.

**Comment by theodorzoulias** [comment]
@stephentoub yes, in this example I do care about the status of the task, but since it is a reference-task I am allowed to inspect the status after the `await`. My point is that I don't expect to get this information by the `await` itself. All that I want from the `await` is to suspend my code until the task is completed, and my [original proposal](https://github.com/dotnet/runtime/issues/22144#issuecomment-1328128898) satisfies this desire. If the `allSourcesCompletion` was a `ValueTask` then my proposal would be inadequate for this case, and I would have to stick with the standard `try`/`catch` approach. Only 10% of my use cases fall in this category (it's a value-task and I am interested about the status).

Maybe it would be more interesting to mention how many of my value-task cases require the status after a non-throwing `await`. My value-task cases are the minority. The majority are reference-tasks. So when I talk about my value-task cases, I am talking about a very small number of cases, numbered in the fingers of one hand. So mentioning percentages is a stress.

Let me say one more thing. I would really like to have the `NoThrow` functionality in a future .NET release, and it's disappointing to see it getting bogged down because the details of the value-task case can't be figured out.

**Comment by stephentoub** [comment]
> Only 10% of my use cases fall in this category (it's a value-task and I am interested about the status).

What are the cases you have where it's a ValueTask and you _don't_ care about the status?  I'm wondering whether your 10% number is really about Task vs ValueTask.

> Let me say one more thing. I would really like to have the NoThrow functionality in a future .NET release, and it's disappointing to see it getting bogged down because the details of the value-task case can't be figured out.

This statement implies the details here don't matter.  I respectfully and significantly disagree with such an assertion.  A NoThrow helper can be copy/pasted into any code base with just a few lines; no one should be blocked by this.  If/when we add it to the core libraries, it's there forever, and it needs to be done "right".

**Comment by theodorzoulias** [comment]
> What are the cases you have where it's a ValueTask and you _don't_ care about the status? I'm wondering whether your 10% number is really about Task vs ValueTask.

See the API usage 2 and 3 in my [original proposal](https://github.com/dotnet/runtime/issues/22144#issuecomment-1328128898). Both involve value-tasks. In the case 2 (`timer.WaitForNextTickAsync`) the assumption is that if the task is not successful, the cause can't be anything else than the cancellation of the token. The token is checked immediately after the `await`, so the status of the token is as good as the status of the task. In the case 3 (`enumerator.MoveNextAsync`) I have included a detailed explanation about why the status of this task is irrelevant.

> This statement implies the details here don't matter. I respectfully and significantly disagree with such an assertion. A NoThrow helper can be copy/pasted into any code base with just a few lines; no one should be blocked by this. If/when we add it to the core libraries, it's there forever, and it needs to be done "right".

Fair enough, I'll wait for good things to come. 😃 I hope that my contribution here helped the forward progress of this API, and I didn't cause unwillingly a regression.

**Comment by stephentoub** [comment]
> In the case 3 (enumerator.MoveNextAsync) I have included a detailed explanation about why the status of this task is irrelevant.

FWIW, I disagree with the assertions in that comment that the exception is irrelevant. First, an action the user's code initiated failed with an arbitrary exception; from my perspective, it shouldn't just be eaten by the implementation because the user happened to stop enumerating (and, yes, while exceptions are discouraged from Dispose, sometimes it's the lesser evil).  Second, the unobserved exception is then likely going to raise the TaskScheduler.UnobservedTaskException event, assuming the operation is actually backed by a task.

I realize this isn't the main point of the discussion, nor do we need to debate the semantics of an API we're not going to ship. I'm simply trying to highlight that it's important we factor in ValueTask.

> I hope that my contribution here helped the forward progress of this API, and I didn't cause unwillingly a regression.

Thanks for the discussion.

**Comment by theodorzoulias** [comment]
@stephentoub some time ago I had a long thought about what to do with a faulted `enumerator.MoveNextAsync` in the `finally` block of an iterator. Eventually I came to the firm conclusion that the exception should be suppressed. Since then I have applied this wisdom to quite a lot of implementations, so now I am a bit biased towards sticking with it. 😃

Regarding the effect of the `NoThrow` on the `TaskScheduler.UnobservedTaskException` event, it was also mentioned in an [earlier comment](https://github.com/dotnet/runtime/issues/22144#issuecomment-760312237) by Reuben Bond. Personally I wouldn't use this event for anything else than logging, and the level of logging would be "Warning" at most. I wouldn't mind seeing a casual warning in my logs about an error that was suppressed by a `NoThrow` awaiting in the `finally` block of my iterator, but I wouldn't miss it either. I would probably be annoyed though if a saw reams of warnings coming from cancellations of `timer.WaitForNextTickAsync` operations!

**Comment by stephentoub** [comment]
I keep coming across places where this would be valuable, and I'd like to get a solution into .NET 8. I think we've been making this too complicated...

We don't need to support ValueTasks, which are problematic because you can't look at them after you await them, and so any model for ValueTask would need to entail returning exception information, which is unfamiliar in the model; we can instead just say "if you need this and you have a ValueTask, call AsTask first". Yes, it _may_ allocate, but the use cases where you need this are more niche, you can check first whether it already completed, and AsTask for something that hasn't yet completed will only allocate if it's backed by an IValueTaskSource, which is itself currently problematic for such configuration because it doesn't provide a way currently to get at the Exception without throwing it.

Further, we don't need to support TResults.  Previous concerns with TResult includes not knowing whether you can actually trust the returned TResult, and any nullability annotations possibly being wrong.  But `Task<TResult>` derives from `Task`, so any support for `Task` "just works" for `Task<TResult>`, and in particular without violating anything about the `TResult`'s nature.

So, if we just expose this support for `Task`, we address the 99% case and we can make that work well.

Thus my proposal is simply this:
```diff
namespace System.Threading.Tasks;

public class Task
{
+    public ConfiguredTaskAwaitable ConfigureAwait(ConfigureAwaitOptions options);
}

+[Flags]
+public enum ConfigureAwaitOptions
+{
     // behaves the same as `await task.ConfigureAwait(false)`
+    None = 0,

     // captures the current SynchronizationContext / TaskScheduler and if present posts the continuation to them
     // behaves the same as `await task` or `await task.ConfigureAwait(true)`
+    ContinueOnCapturedContext = 0x1,

     // awaiting a task that ends as Canceled or Faulted won't throw an exception
     // the consumer can subsequently examine the task to decide how to proceed
+    SuppressExceptions = 0x2,

     // behaves as if the task hasn't yet completed, forcing the continuation to be run asynchronously
     // from the calling thread even if the task has completed.
+    ForceAsynchronousCompletion = 0x4,
+}
```

Notes:
- Whether we need a different return type for this ConfigureAwait overload will depend on implementation and perf testing.  Hopefully we can get away with just using the same type as the one that exists.  If we can't, we'd add a new type with the same shape as `ConfiguredTaskAwaitable` but instead named `ConfiguredWithOptionsTaskAwaitable` (name can obviously be debated).
- As was previously proposed by Pent, we _could_ make this a new awaitable whose GetResult would return the original Task.  That might help some cases, but I'm concerned it could be confusing (especially in a case where you have a `Task<Task>`).
- This _might_ be confusing for someone who actually cares about the current context, such as in client UI, and adds a `.ConfigureAwait` where they didn't previously have one in order to suppress exceptions and neglects to also add ContinueOnCapturedContext.  But to change that, we'd need to invert the name of ContinueOnCapturedContext.

Examples:
```C#
// Wait for a task to complete or a timeout, then see whether the task was successful
await task.WaitAsync(TimeSpan.FromSeconds(10)).ConfigureAwait(ConfigureAwaitOptions.SuppressExceptions);
if (task.IsCompleted) { ... } else { ... }

// Queue the remainder of the method to the thread pool
await Task.CompletedTask.ConfigureAwait(ConfigureAwaitOptions.ForceAsynchronousCompletion);
```

**Comment by bartonjs** [comment]
I can't think (off the top of my head) of any other place where we have a `new` "override" that changes return type where the "base" method also has overloading.  I anticipate confusion... do you anticipate that said confusion will be low?

For `Task` this is easy... you now see two `ConfigureAwait(...)` methods, one takes a boolean, the other takes a Flags enum (the original design finally manifests as reality), they both return a `ConfiguredTaskAwaitable` (unless details say otherwise... but even then they're isomorphic).

For `Task<TResult>`, the `ConfigureAwait(bool)` gets `new`'d to change the return type to a non-voiding awaiter, so `foo = await bar.ConfigureAwait(false);` is legal, but `foo = await bar.ConfigureAwait(ConfigureAwaitOptions.ForceAsynchronousCompletion);` is not.   I understand why functionally the "voiding" awaiter is better, but wonder if the confusion means that the new method should get a different name, and not be part of the logical method group.

I don't have a _better_ name, so I'm unfortunately just doing the "I'm not sure this is good... please solve it."  It might be the best answer and we're OK with it... I'm just sharing my usability/confusion reaction before this comes up in a meeting.

**Comment by stephentoub** [comment]
We have multiple places where overloads return different types, such that changing arguments changes the return type and this can make it no longer assignable to what it was previously. In this case I'm seeing it as a benefit.

**Comment by bartonjs** [comment]
In discussion we feel that it's valid to add this to `Task<T>` as well as `Task`, but that the ConfigureAwait method on `Task<T>` will throw an ArgumentException (or type derived therefrom) if SuppressException is asserted.

We may consider the ValueTask family in later releases.

We should also provide an analyzer that detects the call to `Task<T>.ConfigureAwait` asserting the SuppressException flag, and giving an explicit a priori Warning (category: Usage)

```C#
namespace System.Threading.Tasks;

public partial class Task
{
    public ConfiguredTaskAwaitable ConfigureAwait(ConfigureAwaitOptions options);
}

public partial class Task<TResult>
{
    // throws if ConfigureAwaitOptions.SuppressExceptions is asserted (or any unknown value, of course)
    public new ConfiguredTaskAwaitable<TResult> ConfigureAwait(ConfigureAwaitOptions options);
}

[Flags]
public enum ConfigureAwaitOptions
{
    None = 0,
    ContinueOnCapturedContext = 0x1,
    SuppressExceptions = 0x2,
    ForceYielding = 0x4,
    ForceAsynchronousContinuation = 0x8,
}
```

**Comment by eiriktsarpalis** [comment]
> the ConfigureAwait method on `Task<T>` will throw an ArgumentException

@bartonjs Is `ArgumentException` or `NotSupportedException` more appropriate here?

**Comment by bartonjs** [comment]
Largely I wrote down ArgumentException because it's what @stephentoub said.  But, off the top of my head, of all the Exception types, ArgumentOutOfRangeException seems the most appropriate: "The exception that is thrown when the value of an argument is outside the allowable range of values as defined by the invoked method."

NotSupportedException says "The exception that is thrown when an invoked method is not supported, or when there is an attempt to read, seek, or write to a stream that does not support the invoked functionality."   The "or when..." is really "we should have used InvalidOperationException on those Stream members, but didn't, so let's backform this and say it's OK".  So, NSE is "the method, in its entirety, isn't supported... it probably got here via an interface (or legacy netfx compat)".  The derived PlatformNotSupportedException would make contextual sense if the flag worked on some OSes but not all...

So, really, yeah, it's just AOORE, or some other ArgumentException, because it's "this method doesn't support this argument value, go fix your calling code"

**Comment by stephentoub** [comment]
Fixed in https://github.com/dotnet/runtime/pull/87067

**Comment by TonyValenti** [comment]
@stephentoub Thank you!  I will finally be able to Task.Delay safely!

**Comment by stephentoub** [comment]
> Thank you! I will finally be able to Task.Delay safely!

You're welcome.  By "safely", I assume you mean you were passing it a cancellation token and were having to try/catch(OperationCanceledException) and now you don't have to?

**Comment by davidfowl** [comment]
cc @BrennanConroy

**Comment by stephentoub** [comment]
> cc @BrennanConroy

I'm planning to submit a pr to aspnetcore once it moves to a new enough runtime.

**Comment by TonyValenti** [comment]
Yep!

> > Thank you! I will finally be able to Task.Delay safely!
> 
> 
> 
> You're welcome.  By "safely", I assume you mean you were passing it a cancellation token and were having to try/catch(OperationCanceledException) and now you don't have to?
